barretenberg/cpp/src/barretenberg/relations/toy_avm/generic_lookup_relation.hpp

```diff
@@ -0,0 +1,469 @@
+/**
+ * @file generic_lookup_relation.hpp
+ * @author Rumata888
+ * @brief This file contains the template for the generic lookup that can be specialized to enforce various
+ * lookups (for explanation on how to define them, see "relation_definer.hpp")
+ *
+ * @details Lookup is a mechanism to ensure that a particular value or tuple of values (these can be values of
+ * witnesses, selectors or a function of these) is contained within a particular set. It is a relative of set
+ * permutation, but has a one-to-many relationship beween elements that are being looked up and the table of values they
+ * are being looked up from. In this relation template we use the following terminology:
+ * + READ - the action of looking up the value in the table
+ * + WRITE - the action of adding the value to the lookup table
+ *
+ * TODO(@Rumata888): Talk to Zac why "lookup_read_count" refers to the count of the looked up element in the multiset.
```

Comment:
The relation evaluates the following (with X replaced by a random challenge): 

```
_ 
\  ____1____ -  ____read_count[I]____
/  X + reads[I]     X + writes[I]
¯
```

The reason why we multiple `lookup_read_count` by the write predicate, is because we are *also* multiplying `lookup_read_count` with (1 / X + writes[i]) from the above equation.

Does that make sense?

Reply:
I understood the formula, it was more the semantics that confused me, because we were using read count for the number of times we wrote